src/H5Iint.c

```diff
@@ -8811,505 +8866,162 @@ H5I__discard_mt_id_info(H5I_mt_id_info_t * id_info_ptr)
 
-                }
-            } else {
+                    if ( ! atomic_compare_exchange_strong(&(H5I_mt_g.id_info_fl_shead), 
+                                                          &fl_shead, new_fl_shead) ) {
 
-                // assert(LFHT_FL_NODE_ON_FL == atomic_load(&(fl_next->tag)));
+                        /* the attempt to remove the first item from the free list
+                         * failed.  Update stats and try again.
+                         */
+                        atomic_fetch_add(&(H5I_mt_g.num_id_info_fl_head_update_cols), 1ULL);
 
-                /* attempt to set lfht_ptr->fl_stail to fl_next.  It doesn't
-                 * matter whether we succeed or fail, as if we fail, it
-                 * just means that some other thread beat us to it.
-                 *
-                 * that said, it doesn't hurt to collect stats
-                 */
-                new_fl_stail.ptr = fl_snext.ptr;
-                new_fl_stail.sn  = fl_stail.sn + 1;
-                if ( ! atomic_compare_exchange_strong(&(H5I_mt_g.id_info_fl_stail), &fl_stail, new_fl_stail) ) {
+                    } else {
 
-                    atomic_fetch_add(&(H5I_mt_g.num_id_info_fl_tail_update_cols), 1);
-                }
-            }
-        }
-    }
+                        H5I_mt_id_info_sptr_t null_snext = {NULL, 0ULL};
 
-    /* test to see if both H5I_mt_g.id_info_fl_len and H5I_mt_g.num_id_info_fl_entries_reallocable 
-     * are greater than H5I_mt_g.max_desired_id_info_fl_len.  Check both, as while we want to 
-     * keep the free list length around H5I_mt_g.max_desired_id_info_fl_len, we also want to 
-     * have an ample supply of reallocable free list entries on hand.
-     */
-    assert(atomic_load(&(H5I_mt_g.num_id_info_fl_entries_reallocable)) <= 
-           atomic_load(&(H5I_mt_g.id_info_fl_len)));
+                        /* first has been removed from the free list.  Set id_info_ptr to fl_shead.ptr,
+                         * discard *id_info_ptr, update stats, and exit the loop by setting done to true.
+                         */
+                        id_info_ptr = fl_shead.ptr;
 
-    if ( ( atomic_load(&(H5I_mt_g.id_info_fl_len)) > atomic_load(&(H5I_mt_g.max_desired_id_info_fl_len)) ) &&
-         ( atomic_load(&(H5I_mt_g.num_id_info_fl_entries_reallocable)) > 
-           atomic_load(&(H5I_mt_g.max_desired_id_info_fl_len)) ) ) {
+                        assert(H5I__ID_INFO == id_info_ptr->tag);
+                        assert(atomic_load(&(id_info_ptr->on_fl)));
+                        assert(0 < atomic_load(&(id_info_ptr->serial_num)));
 
-        try_to_free_an_entry = TRUE;
+                        /* the above assert only exists in production builds.  If id_info_ptr->serial_num
+                         * is zero and we get this far, increment H5I_mt_g.num_id_info_fl_head_sn_is_zero.
+                         */
+                        if ( 0 == atomic_load(&(id_info_ptr->serial_num)) ) {
```

Comment:
If an SN of zero indicates that an id info is not on the free list, I think we should assert this is not the case as soon as `fl_shead` is read from `H5I_mt_g`. If the head of the free list is flagged as not being on the free list, something has already gone wrong.

Reply:
See assertion on on_fl.

Reply:
Is there a case where this serial number can be zero here during valid free list operation? If not, I'm not sure I understand the motivation for tracking this with statistics instead of throwing an assertion an early as possible.

Reply:
You are correct that an assertion would not be a bad idea here.  However, the assertion on on_fl is largely redundant.  so I see no point in raising the issue unless there is some other significant issue with the code -- in which case this point can be addressed in passing at minimal cost.

Please keep the cost / benefit tradeoff in mind.